src/Pact/Analyze/Eval/Types.hs

```diff
@@ -0,0 +1,462 @@
+    --
+    , _qeColumnScope   :: Map VarId ()
+    }
+
+mkQueryEnv :: AnalyzeEnv -> AnalyzeState -> AVal -> QueryEnv
+mkQueryEnv env state result = QueryEnv env state result Map.empty Map.empty
+
+newtype Constraints
+  = Constraints { runConstraints :: Symbolic () }
+
+instance Show Constraints where
+  show _ = "<symbolic>"
+
+instance Monoid Constraints where
+  mempty = Constraints (pure ())
+  mappend (Constraints act1) (Constraints act2) = Constraints $ act1 *> act2
+
+data SymbolicCells
+  = SymbolicCells
+    { _scIntValues     :: ColumnMap (SFunArray RowKey Integer)
+    , _scBoolValues    :: ColumnMap (SFunArray RowKey Bool)
+    , _scStringValues  :: ColumnMap (SFunArray RowKey String)
+    , _scDecimalValues :: ColumnMap (SFunArray RowKey Decimal)
+    , _scTimeValues    :: ColumnMap (SFunArray RowKey Time)
+    , _scKsValues      :: ColumnMap (SFunArray RowKey KeySet)
+    -- TODO: opaque blobs
+    }
+  deriving (Show)
+
+-- Implemented by-hand until 8.2, when we have DerivingStrategies
+instance Mergeable SymbolicCells where
+  symbolicMerge force test
+    (SymbolicCells a b c d e f)
+    (SymbolicCells a' b' c' d' e' f')
+    = SymbolicCells (m a a') (m b b') (m c c') (m d d') (m e e') (m f f')
+    where
+      m :: Mergeable a => ColumnMap a -> ColumnMap a -> ColumnMap a
+      m = symbolicMerge force test
+
+-- Checking state that is split before, and merged after, conditionals.
+data LatticeAnalyzeState
+  = LatticeAnalyzeState
+    { _lasSucceeds            :: SBV Bool
+    --
+    -- TODO: instead of having a single boolean here, we should probably use
+    --       finer-grained tracking, so that we can test whether a single
+    --       invariant is being maintained
+    --
+    , _lasMaintainsInvariants :: TableMap (ZipList (Located (SBV Bool)))
+    , _lasTablesRead          :: SFunArray TableName Bool
+    , _lasTablesWritten       :: SFunArray TableName Bool
+    , _lasIntCellDeltas       :: TableMap (ColumnMap (SFunArray RowKey Integer))
+    , _lasDecCellDeltas       :: TableMap (ColumnMap (SFunArray RowKey Decimal))
+    , _lasIntColumnDeltas     :: TableMap (ColumnMap (S Integer))
+    , _lasDecColumnDeltas     :: TableMap (ColumnMap (S Decimal))
+    , _lasTableCells          :: TableMap SymbolicCells
+    , _lasRowsRead            :: TableMap (SFunArray RowKey Integer)
+    , _lasRowsWritten         :: TableMap (SFunArray RowKey Integer)
+    , _lasCellsEnforced       :: TableMap (ColumnMap (SFunArray RowKey Bool))
+    -- We currently maintain cellsWritten only for deciding whether a cell has
+    -- been "invalidated" for the purposes of keyset enforcement. If a keyset
+    -- has been overwritten and *then* enforced, that does not constitute valid
+    -- enforcement of the keyset.
+    , _lasCellsWritten        :: TableMap (ColumnMap (SFunArray RowKey Bool))
+    }
+  deriving (Show)
+
+-- Implemented by-hand until 8.2, when we have DerivingStrategies
+instance Mergeable LatticeAnalyzeState where
```

Comment:
same here. looks like we can now auto-derive this

Reply:
fixed